benchmark_runner/common/clouds/BareMetal/bare_metal_operations.py

```diff
@@ -215,18 +230,64 @@ def run_ocp_installer(self):
         logger.info(f'OpenShift cluster {self._get_installation_version()} version is installed successfully, End time: {datetime.now().strftime(datetime_format)}')
 
     @logger_time_stamp
-    def verify_install_complete(self):
+    def run_ocp_upgrade(self, oc: OC):
         """
-        This method verify that installation complete
+        This method runs ocp upgrade
+        @param oc:
         :return: True if installation success and raise exception if installation failed
         """
+        logger.info(f'Starting OCP upgrade, Start time: {datetime.now().strftime(datetime_format)}')
+        logger.info(f'Stop OCP healthcheck')
+        oc.healthcheck(action='stop')
+        oc.upgrade_ocp(upgrade_ocp_version=self._upgrade_ocp_version)
+        self.verify_upgrade_complete(oc=oc)
+        logger.info(f'Resume OCP healthcheck')
+        oc.healthcheck(action='resume')
+        logger.info(f'Ending OCP upgrade, End time: {datetime.now().strftime(datetime_format)}')
+
+    @logger_time_stamp
+    def verify_install_complete(self):
+        """
+        This method verifies that the installation has completed
+        :return: Returns True if the installation is successful; raises an exception if the installation fails
+        """
         complete = self._wait_for_install_complete()
         if not complete:
             install_log = self._remote_ssh.run_command(self._provision_installer_log)
             raise OCPInstallationFailed(install_log)
         else:
             return True
 
+    def is_cluster_upgraded(self, oc: OC, cnv_version: str, odf_version: str, lso_version: str):
+        """
+        This method checks if cluster/operators upgraded successfully
+        @param oc:
+        @param cnv_version:
+        @param odf_version:
+        @param lso_version:
+        @return:
+        """
+        # check cluster version upgrade
+        if oc.get_upgrade_version() == self._upgrade_ocp_version:
+            lso_upgraded = oc.wait_for_operator_installation(operator='lso', version=lso_version, namespace='openshift-local-storage')
+            odf_upgraded = oc.wait_for_operator_installation(operator='odf', version=odf_version, namespace='openshift-storage')
+            cnv_upgraded = oc.wait_for_operator_installation(operator='cnv', version=cnv_version, namespace='openshift-cnv')
+            # check operators versions upgrade
+            if cnv_upgraded and lso_upgraded and odf_upgraded:
+                return True
+        return False
```

Comment:
Not having all of the operators upgraded (yet) is a different condition than the desired upgrade version being wrong.  The former looks like a condition that may eventually resolve itself.  The latter is not going to be resolved and should probably raise an exception of some kind.

Reply:
Right, add RuntimeError error:
```
        if bare_metal_operations.is_cluster_upgraded(oc, cnv_version=cnv_version, odf_version=odf_version, lso_version=lso_version):
            bare_metal_operations.verify_cluster_is_up(oc)
        else:
            error_message = f'OCP {upgrade_ocp_version} upgrade failed'
            logger.error(error_message)
            raise RuntimeError(error_message)
```

Reply:
OK